.github/workflows/reuse.yaml

```diff
@@ -0,0 +1,134 @@
+          cosign-release: 'v1.4.1'
+
+      - name: Cache Go modules
+        uses: actions/cache@d9747005de0f7240e5d35a68dca96b3f41b8b340 # v1.2.0
+        with:
+          path: ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+
+      - name: Refresh Go report card
+        if: ${{ inputs.tag == 'release'}}
+        uses: creekorful/goreportcard-action@1f35ced8cdac2cba28c9a2f2288a16aacfd507f9 # v1.0
+
+      - name: Log into ghcr.io
+        uses: docker/login-action@master
+        with:
+          registry: ghcr.io
+          username: ${{secrets.registry_username}}
+          password: ${{secrets.registry_password}}
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@4ab11c41e45d028884a99163086648e898eed25 # v1.6.0
+        id: buildx
+        with:
+          install: true
+
+      - name: Set Version
+        if: ${{ inputs.tag == 'release'}}
+        run: |
+          echo "KYVERNO_VERSION=$(git tag --list --sort=-version:refname "v[0-9]*" | head -n 1)" >> $GITHUB_ENV
+
+      - name: Generate SBOM JSON
+        if: ${{ inputs.image_name == 'kyverno' && inputs.tag == 'release'}}
+        uses: CycloneDX/gh-gomod-generate-sbom@v1
+        with:
+          version: v1
+          args: mod -licenses -json -output kyverno-${{ env.KYVERNO_VERSION }}-bom.cdx.json
+          
+      - name: Upload SBOM JSON
+        if: ${{ inputs.image_name == 'kyverno' && inputs.tag == 'release'}}
+        uses: actions/upload-artifact@v2
+        with:
+          name: kyverno-bom-cdx
+          path: kyverno-v*-bom.cdx.json
+
+      - name : Docker images publish
+        run: make ${{inputs.publish_command}}
+
+      - name: get digest
+        id: get-step
+        run: |
+          echo "::set-output name=digest::$(make ${{inputs.digest_command}})"
+
+      - name: Sign image
+        env:
+          COSIGN_EXPERIMENTAL: "true"
+          COSIGN_REPOSITORY: "ghcr.io/kyverno/signatures"
+        run: |          
+          cosign sign \
+          -a "repo=${{ github.repository }}" \
+          -a "workflow=${{ github.workflow }}" \
+          -a "ref=${{ github.sha }}" \
+          ghcr.io/kyverno/${{inputs.image_name}}@sha256:${{ steps.get-step.outputs.digest }}
+
+      - name : Sign SBOM
+        if: ${{ inputs.image_name == 'kyverno' && inputs.tag == 'release'}}
+        env:
+          COSIGN_REPOSITORY: "ghcr.io/kyverno/sbom"
+        run: cosign attach sbom --sbom ./*-bom.cdx.json --type cyclonedx  ghcr.io/kyverno/kyverno:${{env.KYVERNO_VERSION}}
+
+      - name: Trivy Scan Image
+        if: ${{ inputs.tag == 'release'}}
+        uses: aquasecurity/trivy-action@master
+        with: 
+          image-ref: 'ghcr.io/kyverno/${{inputs.image_name}}:${{env.KYVERNO_VERSION}}'
+          format: 'table'
+          exit-code: '1'
+          ignore-unfixed: true
+          vuln-type: 'os,library'
+          severity: 'CRITICAL,HIGH'
```

Comment:
Can we shift this action before we publish any of the images so that if there's a vulnerability, the CI fails before uploading the vulnerable image

Reply:
good point. but if we won't upload then from where will we get the image to scan? 

we are building and pushing the image with the same make command: https://github.com/kyverno/kyverno/blob/dc867e34f8a5eef364f18f66c82aa0c9dc40cdb5/Makefile#L89

Reply:
@Namanl2001 seems like we misplaced another one of my PRs while refactoring. I added the feature to scan the image locally using the filesystem in here https://github.com/kyverno/kyverno/pull/2744 

PTAL and cherry pick the commit if needed

Reply:
> I added the feature to scan the image locally using the filesystem in here #2744

but it's for build.yaml workflow, right? I'm not touching it in any way in this pr.

> The build,yaml will test for the container file system whereas the release.yaml will test the pushed container in GHCR.

this is what you mentioned in this comment: https://github.com/kyverno/kyverno/pull/2744#issuecomment-974907823

also, this comment summarises everything https://github.com/kyverno/kyverno/pull/2744#issuecomment-975148335

Reply:
For this, I would suggest you test this locally by having a vulnerable package and see if Trivy reports it and is scanning the current image before pushing it to our docker registry. Otherwise we'd still push the package and then report the vulnerability which I don't think is the best of choices

Reply:
I agree with @ShubhamPalriwala. Currently we are checking against the output image. Trivy also has a file system mode which checks against the source code which is arguably better than checking the output image since we have more metadata available before the compilation. We should be able to run this check before we build our images and fail the workflow in case a critical vulnerability is detected.

See https://github.com/aquasecurity/trivy-action#using-trivy-to-scan-your-git-repo for an example. 

The reason why we might want to do a scan during the release process even though the build time check runs during PRs is because there might be a significant period of time between PR merge and the release being cut during which a new vulnerabilities may have been discovered.

Reply:
thanks @ShubhamPalriwala and @samj1912 for this sweet suggestion, it will be a nice addition to the kyverno workflows, added this in the latest commit